File: xforms/jvm/src/main/java/org/orbeon/oxf/xforms/processor/handlers/XFormsBaseHandler.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.processor.handlers;

import org.orbeon.oxf.xforms.XFormsConstants;
import org.orbeon.oxf.xforms.XFormsContainingDocument;
import org.orbeon.oxf.xforms.XFormsUtils;
import org.orbeon.oxf.xforms.analysis.ElementAnalysis;
import org.orbeon.oxf.xforms.analysis.controls.AttributeControl;
import org.orbeon.oxf.xforms.control.XFormsControl;
import org.orbeon.oxf.xforms.control.XFormsSingleNodeControl;
import org.orbeon.oxf.xforms.control.controls.XXFormsAttributeControl;
import org.orbeon.oxf.xml.*;
import org.orbeon.xforms.XFormsId;
import org.xml.sax.Attributes;
import org.xml.sax.helpers.AttributesImpl;

/**
 * Base XForms handler used as base class in both the xml and xhtml handlers.
 */
public abstract class XFormsBaseHandler extends ElementHandler {

    private final boolean repeating;
    private final boolean forwarding;

    protected HandlerContext xformsHandlerContext;
    public HandlerContext getXFormsHandlerContext() {
        return xformsHandlerContext;
    }

    protected ElementAnalysis elementAnalysis;
    protected XFormsContainingDocument containingDocument;
    public XFormsContainingDocument getContainingDocument() {
        return containingDocument;
    };

    protected AttributesImpl reusableAttributes = new AttributesImpl();

    protected XFormsBaseHandler(String uri, String localname, String qName, Attributes attributes, Object matched, Object handlerContext, boolean repeating, boolean forwarding) {
        super(uri, localname, qName, attributes, matched, handlerContext);

        this.repeating  = repeating;
        this.forwarding = forwarding;

        // All elements corresponding to XForms controls must cause elementAnalysis to be available
        if (matched instanceof ElementAnalysis)
            this.elementAnalysis = (ElementAnalysis) matched;

        this.xformsHandlerContext = (HandlerContext) handlerContext;
        this.containingDocument = xformsHandlerContext.getContainingDocument();
    }

    public boolean isRepeating() {
        return repeating;
    }
    public boolean isForwarding() {
        return forwarding;
    }

    public boolean isNonRelevant(XFormsControl control) {
        return control == null || ! control.isRelevant();
    }

    public static void handleAccessibilityAttributes(Attributes srcAttributes, AttributesImpl destAttributes) {
        // Handle "tabindex"
        {
            // This is the standard XForms attribute
            String value = srcAttributes.getValue("navindex");
            if (value == null) {
                // Try the the XHTML attribute
                value = srcAttributes.getValue("tabindex");
            }

            if (value != null)
                destAttributes.addAttribute("", "tabindex", "tabindex", XMLReceiverHelper.CDATA, value);
        }
        // Handle "accesskey"
        {
            final String value = srcAttributes.getValue("accesskey");
            if (value != null)
                destAttributes.addAttribute("", "accesskey", "accesskey", XMLReceiverHelper.CDATA, value);
        }
        // Handle "role"
        {
            final String value = srcAttributes.getValue("role");
            if (value != null)
                destAttributes.addAttribute("", "role", "role", XMLReceiverHelper.CDATA, value);
        }
    }

    public static void handleAriaAttributes(boolean required, boolean valid, AttributesImpl destAttributes) {
        if (required)
            destAttributes.addAttribute("", "aria-required", "aria-required", XMLReceiverHelper.CDATA, "true");
        if (! valid)
            destAttributes.addAttribute("", "aria-invalid", "aria-invalid", XMLReceiverHelper.CDATA, "true");
    }

    protected AttributesImpl getIdClassXHTMLAttributes(Attributes elementAttributes, String classes, String effectiveId) {
        return getIdClassXHTMLAttributes(containingDocument, reusableAttributes, elementAttributes, classes, effectiveId);
    }

    public static AttributesImpl getIdClassXHTMLAttributes(XFormsContainingDocument containingDocument, AttributesImpl reusableAttributes, Attributes elementAttributes, String classes, String effectiveId) {
        reusableAttributes.clear();

        // Copy "id"
        if (effectiveId != null) {
            reusableAttributes.addAttribute("", "id", "id", XMLReceiverHelper.CDATA, XFormsUtils.namespaceId(containingDocument, effectiveId));
        }
        // Create "class" attribute if necessary
        {
            if (classes != null && classes.length() > 0) {
                reusableAttributes.addAttribute("", "class", "class", XMLReceiverHelper.CDATA, classes);
            }
        }
        // Copy attributes in the xhtml namespace to no namespace
        for (int i = 0; i < elementAttributes.getLength(); i++) {
            if (XMLConstants.XHTML_NAMESPACE_URI().equals(elementAttributes.getURI(i))) {
                final String name = elementAttributes.getLocalName(i);
                if (!"class".equals(name)) {
                    reusableAttributes.addAttribute("", name, name, XMLReceiverHelper.CDATA, elementAttributes.getValue(i));
                }
            }
        }

        return reusableAttributes;
    }

    public static boolean isStaticReadonly(XFormsControl control) {
        return control != null && control.isStaticReadonly();
    }

    public static String getLHHACId(XFormsContainingDocument containingDocument, String controlEffectiveId, String suffix) {
        // E.g. foo$bar.1-2-3 -> foo$bar$$alert.1-2-3
        return XFormsUtils.namespaceId(containingDocument, XFormsId.appendToEffectiveId(controlEffectiveId, XFormsConstants.LHHAC_SEPARATOR() + suffix));
    }

    protected static Attributes handleAVTsAndIDs(Attributes attributes, String[] refIdAttributeNames, HandlerContext xformsHandlerContext) {
        final String prefixedId = xformsHandlerContext.getPrefixedId(attributes);

        if (prefixedId != null) {

            final XFormsContainingDocument containingDocument = xformsHandlerContext.getContainingDocument();

            final boolean hasAVT = xformsHandlerContext.getPartAnalysis().hasAttributeControl(prefixedId);
            final String effectiveId = xformsHandlerContext.getEffectiveId(attributes);
            boolean found = false;
            if (hasAVT) {
                // This element has at least one AVT so process its attributes

                final int attributesCount = attributes.getLength();
                for (int i = 0; i < attributesCount; i++) {
                    final String attributeValue = attributes.getValue(i);
                    if (XFormsUtils.maybeAVT(attributeValue)) {
                        // This is an AVT most likely
                        found = true;

                        final String attributeLocalName = attributes.getLocalName(i);
                        final String attributeQName = attributes.getQName(i);// use qualified name so we match on "xml:lang"

                        // Control analysis
                        final AttributeControl controlAnalysis = xformsHandlerContext.getPartAnalysis().getAttributeControl(prefixedId, attributeQName);

                        // Get static id of attribute control associated with this particular attribute
                        final String attributeControlStaticId = controlAnalysis.staticId();

                        // Find concrete control
                        final String attributeControlEffectiveId = XFormsId.getRelatedEffectiveId(effectiveId, attributeControlStaticId);
                        final XXFormsAttributeControl attributeControl =
                            (XXFormsAttributeControl) containingDocument.getControlByEffectiveId(attributeControlEffectiveId);

                        // Determine attribute value
                        // NOTE: This also handles dummy images for the xhtml:img/@src case
                        final String effectiveAttributeValue = XXFormsAttributeControl.getExternalValueHandleSrc(attributeControl, controlAnalysis);

                        // Set the value of the attribute
                        attributes = SAXUtils.addOrReplaceAttribute(attributes, attributes.getURI(i),
                                XMLUtils.prefixFromQName(attributeQName), attributeLocalName, effectiveAttributeValue);
                    }
                }

                if (found) {
                    // Update the value of the id attribute
                    attributes = SAXUtils.addOrReplaceAttribute(attributes, "", "", "id", XFormsUtils.namespaceId(containingDocument, effectiveId));
                }
            }

            if (!found) {
                // Id was not replaced as part of AVT processing

                // Update the value of the id attribute
                attributes = SAXUtils.addOrReplaceAttribute(attributes, "", "", "id", XFormsUtils.namespaceId(containingDocument, effectiveId));
            }
        }

        // Check @for or other attribute
        for (String refIdAttributeName : refIdAttributeNames)
        {
            final String forAttribute = attributes.getValue(refIdAttributeName);
            if (forAttribute != null) {
                attributes = SAXUtils.addOrReplaceAttribute(attributes, "", "", refIdAttributeName, xformsHandlerContext.getIdPrefix() + forAttribute + xformsHandlerContext.getIdPostfix());
            }
        }
		return attributes;
	}
}

File: xforms-runtime/shared/src/main/scala/org/orbeon/oxf/xforms/processor/handlers/XFormsBaseHandler.scala
package org.orbeon.oxf.xforms.processor.handlers

import org.orbeon.oxf.xforms.XFormsContainingDocument
import org.orbeon.oxf.xforms.control.XFormsControl
import org.orbeon.oxf.xforms.control.controls.XXFormsAttributeControl
import org.orbeon.oxf.xml.SaxSupport._
import org.orbeon.oxf.xml._
import org.orbeon.xforms.{Constants, XFormsId, XFormsNames}
import org.xml.sax.Attributes
import org.xml.sax.helpers.AttributesImpl


/**
 * Base XForms handler used as base class in both the xml and xhtml handlers.
 */
abstract class XFormsBaseHandler protected (
  uri            : String,
  localname      : String,
  qName          : String,
  attributes     : Attributes,
  handlerContext : HandlerContext,
  val repeating  : Boolean,
  val forwarding : Boolean
) extends
  ElementHandler[HandlerContext](
    uri,
    localname,
    qName,
    attributes,
    handlerContext
  ) {

  protected val containingDocument: XFormsContainingDocument = handlerContext.containingDocument

  // TODO: update to use `repeating` and `forwarding`
  override def isRepeating: Boolean = repeating
  override def isForwarding: Boolean = forwarding

  // TODO: Check callers.
  def isNonRelevant(control: XFormsControl): Boolean =
    (control eq null) || ! control.isRelevant
}

object XFormsBaseHandler {

  def forwardAutocompleteAttribute(srcAttributes: Attributes, localName: String, dstAttributes: AttributesImpl): Unit = {
    if (canHaveAutocompleteAttribute(srcAttributes, localName, dstAttributes)) {
      Option(srcAttributes.getValue("autocomplete")).foreach { autocomplete =>
        dstAttributes.addOrReplace("autocomplete", autocomplete)
      }
    }
  }

  def canHaveAutocompleteAttribute(srcAttributes: Attributes, localName: String, dstAttributes: Attributes): Boolean = {
    val typeOpt =
      Option(dstAttributes.getValue("type")) orElse
      Option(srcAttributes.getValue("type"))

    val supportedInputTypes = Set(
      "color", "date", "datetime-local", "email", "month", "number", "password",
      "range", "search", "tel", "text", "time", "url", "week"
    )

    localName match {
      case "form" | "select" | "textarea"                 => true
      case "input" if typeOpt.forall(supportedInputTypes) => true
      case _                                              => false
    }
  }

  def forwardAccessibilityAttributes(srcAttributes: Attributes, destAttributes: AttributesImpl): Unit = {

    // Handle "tabindex"
    locally {
      // This is the standard XForms attribute
      var value = srcAttributes.getValue("navindex")
      if (value == null) {
        // Try the the XHTML attribute
        value = srcAttributes.getValue(XFormsNames.TABINDEX_QNAME.localName)
      }
      if (value != null)
        destAttributes.addOrReplace(XFormsNames.TABINDEX_QNAME, value)
    }

    // Handle "accesskey"
    locally {
      val value = srcAttributes.getValue(XFormsNames.ACCESSKEY_QNAME.localName)
      if (value ne null)
        destAttributes.addOrReplace(XFormsNames.ACCESSKEY_QNAME, value)
    }

    // Handle "role"
    locally {
      val value = srcAttributes.getValue(XFormsNames.ROLE_QNAME.localName)
      if (value ne null)
        destAttributes.addOrReplace(XFormsNames.ROLE_QNAME, value)
    }
  }

  def handleAriaAttributes(required: Boolean, valid: Boolean, visited: Boolean, destAttributes: AttributesImpl): Unit = {
    if (required)
      destAttributes.addOrReplace("aria-required", "true")
    if (! valid && visited)
      destAttributes.addOrReplace("aria-invalid", "true")
  }

  def getIdClassXHTMLAttributes(
    containingDocument : XFormsContainingDocument,
    elementAttributes  : Attributes,
    classes            : String,
    effectiveIdOpt     : Option[String]
  ): AttributesImpl = {

    val atts = new AttributesImpl

    // Copy "id"
    effectiveIdOpt foreach { effectiveId =>
      atts.addOrReplace(XFormsNames.ID_QNAME, containingDocument.namespaceId(effectiveId))
    }

    // Create "class" attribute if necessary
    if (classes != null && classes.nonEmpty)
      atts.addOrReplace(XFormsNames.CLASS_QNAME, classes)

    // Copy attributes in the xhtml namespace to no namespace
    for (i <- 0 until elementAttributes.getLength) {
      if (XMLConstants.XHTML_NAMESPACE_URI == elementAttributes.getURI(i)) {
        val name = elementAttributes.getLocalName(i)
        if (name != "class")
          atts.addOrReplace(name, elementAttributes.getValue(i))
      }
    }
    atts
  }

  def isStaticReadonly(control: XFormsControl): Boolean =
    control != null && control.isStaticReadonly

  // E.g. `foo≡bar⊙2` -> `foo≡bar≡≡a⊙2`
  def getLHHACIdWithNs(
    containingDocument : XFormsContainingDocument,
    controlEffectiveId : String,
    suffix             : String
  ): String =
    containingDocument.namespaceId(XFormsId.appendToEffectiveId(controlEffectiveId, Constants.LhhacSeparator + suffix))

  def handleAVTsAndIDs(
    _attributes          : Attributes,
    refIdAttributeNames  : Array[String],
    xformsHandlerContext : HandlerContext
  ): Attributes = {

    var attributes = _attributes

    val prefixedId = xformsHandlerContext.getPrefixedId(attributes)
    if (prefixedId ne null) {

      val containingDocument = xformsHandlerContext.containingDocument
      val hasAVT             = xformsHandlerContext.getPartAnalysis.hasAttributeControl(prefixedId)
      val effectiveId        = xformsHandlerContext.getEffectiveId(attributes)

      var found = false
      if (hasAVT) {
        // This element has at least one AVT so process its attributes
        val attributesCount = attributes.getLength
        for (i <- 0 until attributesCount) {
          val attributeValue = attributes.getValue(i)
          if (XMLUtils.maybeAVT(attributeValue)) {
            // This is an AVT most likely
            found = true
            val attributeLocalName = attributes.getLocalName(i)
            val attributeQName = attributes.getQName(i) // use qualified name so we match on "xml:lang"

            // Control analysis
            val controlAnalysis = xformsHandlerContext.getPartAnalysis.getAttributeControl(prefixedId, attributeQName)

            // Get static id of attribute control associated with this particular attribute
            val attributeControlStaticId = controlAnalysis.staticId

            // Find concrete control
            val attributeControlEffectiveId = XFormsId.getRelatedEffectiveId(effectiveId, attributeControlStaticId)
            val attributeControl = containingDocument.getControlByEffectiveId(attributeControlEffectiveId).asInstanceOf[XXFormsAttributeControl]

            // Determine attribute value
            // NOTE: This also handles dummy images for the xhtml:img/@src case
            val effectiveAttributeValue = XXFormsAttributeControl.getExternalValueHandleSrc(attributeControl, controlAnalysis)
            // Set the value of the attribute
            attributes = XMLReceiverSupport.addOrReplaceAttribute(attributes, attributes.getURI(i), XMLUtils.prefixFromQName(attributeQName), attributeLocalName, effectiveAttributeValue)
          }
        }
        if (found) // update the value of the id attribute
          attributes = XMLReceiverSupport.addOrReplaceAttribute(attributes, "", "", "id", containingDocument.namespaceId(effectiveId))
      }
      if (! found) // id was not replaced as part of AVT processing
        attributes = XMLReceiverSupport.addOrReplaceAttribute(attributes, "", "", "id", containingDocument.namespaceId(effectiveId))
    }
    // Check `@for` or other attribute
    for (refIdAttributeName <- refIdAttributeNames) {
      val forAttribute = attributes.getValue(refIdAttributeName)
      if (forAttribute != null)
        attributes = XMLReceiverSupport.addOrReplaceAttribute(attributes, "", "", refIdAttributeName, xformsHandlerContext.getIdPrefix + forAttribute + xformsHandlerContext.getIdPostfix)
    }
    attributes
  }
}
